<!DOCTYPE html>
<!--[if lt IE 7]> <html class="lt-ie9 lt-ie8 lt-ie7"> <![endif]-->
<!--[if IE 7]>    <html class="lt-ie9 lt-ie8"> <![endif]-->
<!--[if IE 8]>    <html class="lt-ie9"> <![endif]-->
<!--[if gt IE 8]><!--> <html class=""> <!--<![endif]-->
    <head>
        <title>fathmm - About Our Software and Server</title>

        <meta charset="utf-8">
        
        <meta name="viewport" content="width=device-width, initial-scale=1.0">
        <meta name="description" content="About Our Software and Server">
        <meta name="keywords" content="fathmm">
        <meta name="author" content="Hashem A. Shihab">
        
        <meta http-equiv="cache-control" content="max-age=0" />
        <meta http-equiv="cache-control" content="no-cache" />
        <meta http-equiv="expires" content="0" />
        <meta http-equiv="expires" content="Tue, 01 Jan 1980 1:00:00 GMT" />
        <meta http-equiv="pragma" content="no-cache" />
        
        <link rel="icon" type="image/png" href="./img/favicon.ico">
        
        <!-- HTML5 "Upscaling" -->
        <!--[if lt IE 9]>
            <script src="http://html5shiv.googlecode.com/svn/trunk/html5.js"></script>
        <![endif]-->

        <link href="./css/bootstrap.css" rel="stylesheet">
        <link href="./css/bootstrap-responsive.css" rel="stylesheet">
        <link href="./css/fathmm.css" rel="stylesheet" type="text/css">
        
        <script type="text/javascript" src="./js/bootstrap.min.js"></script>
        <script type="text/javascript" src="./js/jquery.min.js"></script>
        
        <script type="text/javascript">
          var _gaq = _gaq || [];
          _gaq.push(['_setAccount', 'UA-29568329-3']);
          _gaq.push(['_trackPageview']);

          (function() {
            var ga = document.createElement('script'); ga.type = 'text/javascript'; ga.async = true;
            ga.src = ('https:' == document.location.protocol ? 'https://ssl' : 'http://www') + '.google-analytics.com/ga.js';
            var s = document.getElementsByTagName('script')[0]; s.parentNode.insertBefore(ga, s);
          })();
        </script>
    </head>

    <body>
        <a id="top"></a>
        
        <div class="container">
            <div class="navbar navbar-fixed-top">
                <div class="navbar-inner">
                    <div class="container">
                        <a class="btn btn-navbar" data-toggle="collapse" data-target=".nav-collapse">
                            <span class="icon-bar"></span>
                            <span class="icon-bar"></span>
                            <span class="icon-bar"></span>
                        </a>
                        <a class="brand" href="">fathmm</a>
                        <div class="nav-collapse collapse">
                            <ul class="nav">
                                <li><a href="./index.html">Home</a></li>
                                <li><a href="./about.html">About</a></li>
                                <li><a href="https://github.com/HAShihab/fathmm">Download</a></li>
                            </ul>
                            <form class="navbar-form pull-right" action="./cgi-bin/results.cgi" method="post" enctype="multipart/form-data">
                                <input class="span3" type="text" id="session" name="session" placeholder="Enter Your Job/Session Identifier">
                                <button type="submit" class="btn">Fetch Results</button>
                            </form>
                        </div><!--/.nav-collapse -->
                    </div>
                </div>
            </div>
            
            <div class="hero-unit">
                <h2>About Our Software and Server</h2>
            </div>
            

            <div class="tab-content">
                <div class="tab-pane active">
                    
                    <div class="row">
                        <div class="span3" id="navparent">
                            <ul class="nav nav-list navex">
                                <li><a href="#overview"><i class="icon-chevron-right"></i> Overview</a></li>
                                <li><a href="#format"><i class="icon-chevron-right"></i> Input Format</a></li>
                                <li><a href="#format_batch"><i class="icon-chevron-right"></i> Batch Submission</a></li>
                                <li><a href="#vcf"><i class="icon-chevron-right"></i> Annotating VCF's</a></li>
                            </ul>
                        </div>
                        <div class="span9">
                            
                            <section id="overview">
                                <h1>Overview:</h1>
                                <br />
                                <p>
                                    The Functional Analysis through Hidden Markov Models (fathmm) software and server is is capable of <b>predicting the functional 
                                    effects of protein missense mutations</b> by combining sequence conservation within hidden Markov models (HMMs), representing the 
                                    alignment of homologous (orthologous and/or paralogous) sequences and conserved protein domains, with "pathogenicity weights", representing the overall tolerance
                                    of the corresponding model to mutations.
                                    <br /><br />
                                    For more information, please refer to the following publications:
                                    <br /><br />
                                    Shihab HA, Gough J, Cooper DN, Stenson PD, Barker GLA, Edwards KJ, Day INM, Gaunt, TR. (2013). Predicting the Functional, Molecular and Phenotypic
                                    Consequences of Amino Acid Substitutions using Hidden Markov Models. <i>Hum. Mutat.</i>, <b>34</b>:57-65
                                    <a href="http://www.ncbi.nlm.nih.gov/pubmed/23033316"><img src="./img/pubmed.png" alt="fathmm - Main Paper"></a>
                                    <br /><br />
                                    Shihab HA, Gough J, Cooper DN, Day INM, Gaunt, TR. (2013). Predicting the Functional Consequences of Cancer-Associated Amino Acid Substitutions. 
                                    <i>Bioinformatics</i> <b>29</b>:1504-1510.
                                    <a href="http://www.ncbi.nlm.nih.gov/pubmed/23620363"><img src="./img/pubmed.png" alt="fathmm - Cancer Paper"></a>
                                </p>
                                <br />
                                <a href='./about.html#top'>Back to Top ...</a>
                            </section>
                            <br />
                            
                            <section id="format">
                                <h1>Input Format:</h1>
                                <br />
                                <p>
                                Our software accepts one of the following formats (see <a href="./about.html#vcf">here</a> for annotating VCF files):
                                
                                <br /><br />
                                <ul>
                                    <li>
                                        <code>&lt;protein&gt; &lt;substitution&gt;</code>
                                    </li>
                                    <li>
                                        <code>dbSNP rs identifiers</code>
                                    </li>
                                </ul>
                                <br />
                                In the above, <code>&lt;protein&gt;</code> is the protein identifier and <code>&lt;substitution&gt;</code> is the amino acid substitution in the conventional one 
                                letter format.  At present, <b>our server accepts SwissProt/TrEMBL, RefSeq and Ensembl protein identifiers</b>, e.g.:
                                <br /><br />
<pre>
P43026 L441P
</pre>
                                or:
                                <br /><br />  
<pre>
rs137854462
</pre>
                                </p>
                                <br />
                                <a href='./about.html#top'>Back to Top ...</a>
                            </section>
                            <br />

                            <section id="format_batch">
                                <h1>Batch Submission:</h1>
                                <br />
                                <p>
                                <b>
                                It is possible to submit multiple amino acid substitutions as a 'Batch Submission' via our server</b>.  Here, all amino acid substitutions for a protein can be 
                                entered on a single line and should be separated by a comma, e.g:
                                <br /><br />
<pre>
P43026 L441P
ENSP00000325527 N548I,E1073K,C2307S 
</pre>
                                <br />
                                <b>Note: this option is not available when analysing dbSNP rs identifiers</b>.
                                </p>
                                <br />
                                <a href='./about.html#top'>Back to Top ...</a>
                            </section>
                            <br />
                            
                            <section id="vcf">
                                <h1>VCF Annotation:</h1>
                                <br />
                                <p>
                                Unfortunately, due to disk space constraints, we are unable to annotate Variant Call Format (VCF) files on your behalf.  However, the consequences of all VCF variants 
                                can be derived using the <b><a href="http://www.ensembl.org/info/docs/variation/vep/index.html">Ensembl Variant Effect Predictor (VEP)</a></b>.
                                Once annotated, the following script <b><a href="./parseVCF.py">(available here)</a></b> is capable of parsing these annotations and will provide you with a list of protein 
                                consequences which can then be used as input into our server/software.  
                                <br /><br />
                                Additional help on using our script is available by typing the following command:
                                <br /><br />
                                <pre>python parseVCF.py --help</pre>         
                                </p> 
                                <br />
                                <a href='./about.html#top'>Back to Top ...</a>
                            </section>
                            
                        </div>
                    </div>
                    
                    
                </div>
            </div>
            
            <hr>
            <footer>
                <p>
                    If you have found this resource useful, please cite the following work:
                    <br />
                    <a href="http://www.ncbi.nlm.nih.gov/pubmed/23033316">Shihab HA, Gough J, Cooper DN, Stenson PD, Barker GLA, Edwards KJ, Day INM, Gaunt, TR. (2013). Predicting the Functional, Molecular and Phenotypic Consequences of Amino Acid Substitutions using Hidden Markov Models. Hum. Mutat., <b>34</b>, 57-65 </a>
                </p>
                <p>
                    We welcome any comments and/or suggestions that you may have regarding our software and server - please send an email directly to fathmm@biocompute.org.uk
                </p>
            </footer>
        </div>
        
        <script type="text/javascript" src="./js/bootstrap.min.js"></script>
        <script type="text/javascript" src="./js/jquery.min.js"></script>
        
    </body>
</html>
